src/Microsoft.ML.Tokenizers/Model/SentencePieceBpe.cs

```diff
@@ -0,0 +1,952 @@
+                // escape the dummy prefix if needed.
+                AppendSpan(ref sb, AddDummyPrefix && !TreatWhitespaceAsSuffix && token.Length > 0 && token[0] == LlamaNormalizer.DummyPrefix ?
+                                    token.AsSpan(1) :
+                                    token.AsSpan());
+            }
+
+            char[]? charPoolArray = null;
+
+            while (enumerator.MoveNext())
+            {
+                if (enumerator.Current < _byteCodeToIdOffset)
+                {
+                    if (bytesCount >= 1)
+                    {
+                        FlushBytes(ref bytesCount, ref bytesPoolArray, ref charPoolArray, ref sb);
+                    }
+                    continue;
+                }
+
+                if (enumerator.Current <= _maxByteId)
+                {
+                    if (bytesCount >= 1)
+                    {
+                        Debug.Assert(bytesPoolArray is not null);
+
+                        if (bytesCount >= bytesPoolArray!.Length)
+                        {
+                            Helpers.ArrayPoolGrow(ref bytesPoolArray, bytesCount * 2);
+                        }
+
+                        bytesPoolArray![bytesCount++] = (byte)(enumerator.Current - _byteCodeToIdOffset);
+                    }
+                    else
+                    {
+                        EncodeByte(enumerator.Current, _oneByteUtf8EncodingMaxId, _byteCodeToIdOffset, ref bytesCount, ref bytesPoolArray, ref sb);
+                    }
+                }
+                else
+                {
+                    if (bytesCount >= 1)
+                    {
+                        FlushBytes(ref bytesCount, ref bytesPoolArray, ref charPoolArray, ref sb);
+                    }
+
+                    if (_vocabReverse.TryGetValue(enumerator.Current, out string? token))
+                    {
+                        AppendSpan(ref sb, token.AsSpan());
+                    }
+                }
+            }
+
+            if (bytesCount >= 1)
+            {
+                FlushBytes(ref bytesCount, ref bytesPoolArray, ref charPoolArray, ref sb);
+            }
+
+            if (AddDummyPrefix && TreatWhitespaceAsSuffix && sb.Length > 0 && sb[sb.Length - 1] == LlamaNormalizer.DummyPrefix)
+            {
+                sb.RemoveLastChar();
+            }
+
+            if (bytesPoolArray is not null)
+            {
+                ArrayPool<byte>.Shared.Return(bytesPoolArray);
+            }
+
+            if (charPoolArray is not null)
+            {
+                ArrayPool<char>.Shared.Return(charPoolArray);
+            }
+
+            return sb.ToString();
+
+            static void AppendSpan(ref ValueStringBuilder sb, ReadOnlySpan<char> span)
+            {
+                for (int i = 0; i < span.Length; i++)
+                {
+                    char c = span[i];
+                    sb.Append(c == LlamaNormalizer.DummyPrefix ? ' ' : c);
+                }
```

Comment:
Is DummyPrefix common? It'd be cheaper to do something like this:
```C#
int pos = sb.Length;
sb.Append(span);
sb.AsSpan(pos).Replace(LlamaNormalizer.DummyPrefix, ' ');
```
That Replace method is new in .NET 8, so for the netstandard build, you could just have a polyfill extension method on span that does the same for loop you're currently doing, e.g.
```C#
internal static void Replace(this Span<char> span, char oldValue, char newValue)
{
    for (int i = 0; i < span.Length; i++)
        if (span[i] == oldValue)
            span[i] = newValue;
}
```

Reply:
interesting! DummyPrefix is common as it replaces all spaces in the sentence. But I am seeing your idea still hold. I'll go with your suggestion but will call Replace at the end before calling sb.ToString(). Thanks!